File: udp_client.c
/* udp_client.c */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>

#define BUF_SIZE 1024

void error_handling(char *message);

int main(int argc, char *argv[])
{
    int sock;
    char message[BUF_SIZE];
    int str_len;
    socklen_t adr_sz;
    
    // serv_adr: 메시지를 "보낼" 서버의 주소
    // from_adr: 메시지를 "받은" 서버의 주소 (확인용)
    struct sockaddr_in serv_adr, from_adr;

    if (argc != 3) {
        printf("Usage : %s <IP> <port>\n", argv[0]);
        exit(1);
    }

    // 1. 소켓 생성 (socket) - SOCK_DGRAM (UDP)
    sock = socket(PF_INET, SOCK_DGRAM, 0);   
    if (sock == -1)
        error_handling("socket() error");

    // "받는 사람" (서버)의 주소 구조체 설정
    memset(&serv_adr, 0, sizeof(serv_adr));
    serv_adr.sin_family = AF_INET;
    serv_adr.sin_addr.s_addr = inet_addr(argv[1]); // 서버 IP
    serv_adr.sin_port = htons(atoi(argv[2]));      // 서버 Port
    
    // 2. connect() 과정이 없습니다!

    while (1) 
    {
        fputs("Input message (Q to quit): ", stdout);
        fgets(message, sizeof(message), stdin);
        
        if (strcmp(message, "q\n") == 0 || strcmp(message, "Q\n") == 0)
            break;

        // 3. 데이터 송신 (sendto)
        // "serv_adr" (서버 주소)로 "message"를 보냄
        sendto(sock, message, strlen(message), 0, 
               (struct sockaddr*)&serv_adr, sizeof(serv_adr));
        
        adr_sz = sizeof(from_adr);
        
        // 4. 데이터 수신 (recvfrom)
        // 서버로부터 온 에코 메시지를 받음
        str_len = recvfrom(sock, message, BUF_SIZE, 0, 
                           (struct sockaddr*)&from_adr, &adr_sz);
        
        message[str_len] = 0; // 문자열 종료 처리
        printf("Message from server: %s", message);
    }

    // 5. 소켓 닫기 (close)
    close(sock);
    return 0;
}

void error_handling(char *message)
{
    perror(message);
    exit(1);
}

File: udp_client.py
# udp_client.py
import socket

BUF_SIZE = 1024
HOST = '127.0.0.1'  # 서버 IP
PORT = 9999         # 서버 Port

# 1. 소켓 생성 (socket)
with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
    
    # 2. connect() 과정이 없음
    
    while True:
        try:
            message = input("서버로 보낼 메시지 (종료: 'exit'): ")
            if message.lower() == 'exit':
                break

            # 3. 데이터 송신 (sendto)
            # C의 sendto()와 동일
            # 메시지를 바이트로 인코딩하여 (HOST, PORT) 주소로 보냄
            # 데이터를 보낼 때, 매번 받는 사람 주소 명시
            s.sendto(message.encode('utf-8'), (HOST, PORT))

            # 4. 데이터 수신 (recvfrom)
            # 서버로부터 에코(응답) 데이터를 받음
            # 데이터를 받을 때, 보낸 사람 주소가 항상 반환
            data, addr = s.recvfrom(BUF_SIZE)
            
            print(f"서버({addr})로부터 에코: {data.decode('utf-8')}")

        except KeyboardInterrupt:
            print("\n사용자에 의해 종료됩니다.")
            break

print("클라이언트를 종료합니다.")